### PR TITLE
👨‍🏭 move more logic out of `create raw model`

### PR DESCRIPTION
**Pull request by OsmondJiang**
to make out output model more like raw model prepared for template

### CODE REVIEW

File: docs/specs/legacy.yml [hunk 1]
```diff
@@ -134,35 +146,48 @@ outputs:
...
+        "_op_gitContributorInformation": {
+          "update_at": "10/30/2018",
+          "updated_at_date_time": "2018-10-30T00:00:00Z",
+          "author": {
+            "name": "charlie",
+            "profile_url": "https://github.com/charlie",
+            "display_name": "Charlie",
+            "id": "3"
+          },
+          "!contributors": null
```
Comment (OsmondJiang) [inline docs/specs/legacy.yml]: that's a break change, will break template logic, any idea? @yufeih  @live1206 

image

File: src/docfx/build/page/BuildPage.cs [hunk 2]
```diff
@@ -33,13 +33,17 @@ internal static class BuildPage
...
+            model.EnableLocSxs = file.Docset.Config.Localization.Bilingual;
+            model.SiteName = file.Docset.Config.SiteName;
 
             (model.DocumentId, model.DocumentVersionIndependentId) = file.Docset.Redirections.TryGetDocumentId(file, out var docId) ? docId : file.Id;
             (model.ContentGitUrl, model.OriginalContentGitUrl, model.OriginalContentGitUrlTemplate, model.Gitcommit) = context.ContributionProvider.GetGitUrls(file);
 
             List<Error> contributorErrors;
-            (contributorErrors, model.AuthorInfo, model.Contributors, model.UpdatedAt) = await context.ContributionProvider.GetAuthorAndContributors(file, model.Author);
+            (contributorErrors, model.ContributionInfo) = await context.ContributionProvider.GetContributionInfo(file, model.Author);
+            model.Author = model.ContributionInfo?.Author?.Name == null ? null : new SourceInfo<string>(model.ContributionInfo?.Author?.Name, model.Author);
```
Comment (OsmondJiang) [inline src/docfx/build/page/BuildPage.cs]: that's a workaround, seems SourceInfo<T> doesn't follow our default contract resolver (value null -> disappear)

File: src/docfx/build/page/OutputModel.cs [hunk 3]
```diff
@@ -31,11 +29,10 @@ public class OutputModel : InputMetadata
 
         public string DocumentVersionIndependentId { get; set; }
 
-        public Contributor AuthorInfo { get; set; }
-
-        public List<Contributor> Contributors { get; set; }
+        [JsonProperty("_op_gitContributorInformation"/*legacy name*/)]
+        public ContributionInfo ContributionInfo { get; set; }
```
Comment (yufeih) [inline src/docfx/build/page/OutputModel.cs]: 😱😱😱 As much as I hate it, it is a good change.

File: src/docfx/template/TemplateEngine.cs [hunk 4]
```diff
@@ -188,47 +185,13 @@ public JObject CreateRawMetadata(OutputModel pageModel, Document file)
...
-                }
-            }
-            rawMetadata.Remove("author_info");
-
-            if (!string.IsNullOrEmpty(pageModel.AuthorInfo?.Name))
-                rawMetadata["author"] = pageModel.AuthorInfo?.Name;
-            rawMetadata.Remove("contributors");
-
-            if (pageModel.UpdatedAt != default)
-                rawMetadata["updated_at"] = pageModel.UpdatedAt.ToString("yyyy-MM-dd hh:mm tt");
```
Comment (OsmondJiang) [inline src/docfx/template/TemplateEngine.cs]: this will generate different string value based on the time zone

File: docs/specs/legacy.yml (hunk 1 above)
Comment (OsmondJiang) [reply docs/specs/legacy.yml]: https://ceapex.visualstudio.com/Engineering/_git/docs-ui/pullrequest/10814?_a=overview has been created to fix it in template.

File: src/docfx/build/page/BuildPage.cs (hunk 2 above)
Comment (live1206) [reply src/docfx/build/page/BuildPage.cs]: fixing in #4382

Comment (live1206) [reply src/docfx/build/page/BuildPage.cs]: The null check can be removed now if merge from v3

File: src/docfx/build/page/OutputModel.cs [hunk 5]
```diff
@@ -31,11 +29,10 @@ public class OutputModel : InputMetadata
 
         public string DocumentVersionIndependentId { get; set; }
 
-        public Contributor AuthorInfo { get; set; }
+        [JsonProperty("_op_gitContributorInformation"/*legacy name*/)]
```
Comment (yufeih) [inline src/docfx/build/page/OutputModel.cs]:
```suggestion
        [JsonProperty("_op_gitContributorInformation")]
```

File: src/docfx/build/metadata/ContributionProvider.cs [hunk 6]
```diff
@@ -28,7 +28,87 @@ public ContributionProvider(Docset docset, GitHubUserCache gitHubUserCache, GitC
             LoadCommitsTime(docset);
         }
 
-        public async Task<(List<Error> error, Contributor author, List<Contributor> contributors, DateTime updatedAt)> GetAuthorAndContributors(
+        public async Task<(List<Error> errors, ContributionInfo contributionInfo)> GetContributionInfo(Document document, SourceInfo<string> authorName)
+        {
+            var (errors, author, contributors, updatedAt) = await GetAuthorAndContributors(document, authorName);
+
+            return (errors, updatedAt != default
```
Comment (yufeih) [inline src/docfx/build/metadata/ContributionProvider.cs]: nit: I would directly rename `GetAuthorAndContributors` to `GetContributionInfo` and make it return `(Errors, ContributionInfo)`, there is no need to create a wrapper function just to convert the return value type.

File: docs/specs/git.yml [hunk 7]
```diff
@@ -30,13 +30,16 @@ inputs:
...
     {
-      "author_info": 
-        { "name": "charlie", "profile_url": "https://github.com/charlie", "display_name": "Charlie", "id": "3" },
-      "contributors": [
-        { "name": "bob", "profile_url": "https://github.com/bob", "display_name": "Bob", "id": "2" },
-        { "name": "alice", "profile_url": "https://github.com/alice", "display_name": "Alice", "id": "1" }
-      ],
-      "updated_at": "2018-01-01T00:00:00.0000000"
+      "_op_gitContributorInformation":
+      {
```
Comment (yufeih) [inline docs/specs/git.yml]: also check `_op_gitContributorInformation.updated_at` and `_op_gitContributorInformation.updated_at_datetime` in at least one non-legacy test.

File: src/docfx/build/metadata/ContributionProvider.cs [hunk 8]
```diff
@@ -28,7 +28,87 @@ public ContributionProvider(Docset docset, GitHubUserCache gitHubUserCache, GitC
...
+                ? new ContributionInfo
+                {
+                    Contributors = contributors ?? new List<Contributor>(),
+                    /*it's a workaround to produce stable `update_at` in different platform for en-us content*/
```
Comment (yufeih) [inline src/docfx/build/metadata/ContributionProvider.cs]:
```suggestion
                    // it's a workaround to produce stable `update_at` in different platform for en-us content
```

**Comment by OsmondJiang** [comment]
this PR need wait for https://ceapex.visualstudio.com/Engineering/_git/docs-ui/pullrequest/10814?_a=overview to be released

**Comment by OsmondJiang** [comment]
image
it's released!

File: src/docfx/build/metadata/ContributionProvider.cs [hunk 9]
```diff
@@ -80,7 +86,13 @@ public ContributionProvider(Docset docset, GitHubUserCache gitHubUserCache, GitC
                 contributors.RemoveAll(c => c.Id == author.Id);
             }
 
-            return (errors, author, contributors, updatedDateTime);
+            if (contributionInfo != null)
+            {
+                contributionInfo.Author = author;
+                contributionInfo.Contributors = contributors;
```
Comment (mingwli) [inline src/docfx/build/metadata/ContributionProvider.cs]: >contributionInfo.Contributors = contributors; [](start = 16, length = 45)

is this needed?

Comment (OsmondJiang) [reply src/docfx/build/metadata/ContributionProvider.cs]: logically we want this

**Pull request merged**